test/metabase/models/dashboard_test.clj

```diff
@@ -634,6 +634,62 @@
+                           :model/Card      {will-be-archived-card :id} {}
+                           :model/Card      {will-be-deleted-card :id}  {}
+                           :model/Card      {unchanged-card       :id}  {}]
+    ;; 0. create a dashboard
+    (create-dashboard-revision! dashboard-id true)
+    ;; 1. add 3 cards and 1 text card
+    (t2/insert-returning-pks! :model/DashboardCard
+                              [{:dashboard_id     dashboard-id
+                                :dashboard_tab_id nil
+                                :card_id          will-be-archived-card
+                                :row              0
+                                :col              0
+                                :size_x           4
+                                :size_y           4}
+                               {:dashboard_id     dashboard-id
+                                :dashboard_tab_id nil
+                                :card_id          will-be-deleted-card
+                                :row              4
+                                :col              4
+                                :size_x           4
+                                :size_y           4}
+                               {:dashboard_id     dashboard-id
+                                :dashboard_tab_id nil
+                                :card_id          unchanged-card
+                                :row              0
+                                :col              0
+                                :size_x           4
+                                :size_y           4}
+                               {:dashboard_id           dashboard-id
+                                :dashboard_tab_id       nil
+                                :row                    4
+                                :col                    4
+                                :size_x                 4
+                                :size_y                 4
+                                :visualization_settings {:text "Metabase"}}])
+    (create-dashboard-revision! dashboard-id false)
+
+    ;; 2. delete all the dashcards
+    (t2/delete! :model/DashboardCard :dashboard_id dashboard-id)
+    (create-dashboard-revision! dashboard-id false)
+
+    ;; cards can be deleted if its database is deleted
+    (t2/delete! :model/Card will-be-deleted-card)
+    (t2/update! :model/Card :id will-be-archived-card {:archived true})
+
+    (testing "revert should not includes archived or deleted card ids (#34884)"
+      (revert-to-previous-revision Dashboard dashboard-id 2)
+      (is (=? #{{:card_id                unchanged-card
+                 :visualization_settings {}}
+                {:card_id                nil
+                 :visualization_settings {:text "Metabase"}}}
+              (t2/select-fn-set #(select-keys % [:card_id :visualization_settings])
+                                :model/DashboardCard :dashboard_id dashboard-id))))))
```

Comment:
This test is much more specific than necessary, because it includes details on visualization_settings. All we need to
test is that the set of dashboard cards excludes dashboards for archived or deleted cards. To make tests less brittle, you should try to write tests that have as
little specific details
as possible. Here's an idea to make this test less specific while still matching the goal of the test:
```
(is (= #{unchanged-dashcard-id-1 unchanged-dashcard-id-2}
       (t2/select-pk-set :model/DashboardCard :dashboard_id dashboard-id)))
```
(assuming you use with-temp to create the dashcards and you have the dashcard IDs bounded already)

Reply:
It's implicit in the test but it was my intention to include a text card.

it's used to make sure `remove-invalid-dashcards` do not remove text cards by accident.

I'm not sure whether we have tests that cover this or not but it's nice to have it here.
Would a comment clear it up?

Reply:
A comment would clear up why the text card is included, but that's not my point here

I'm thinking of ways to minimize the number of changes you'd have to make to this test if unrelated parts of the codes change, such as the way text cards are represented.

We probably don't have a `create-mock-text-card` function (maybe we should?), so we can't get rid of all the details when creating the text dashcard. But we can at least avoid littering the `(is ...)` expression with these details because we should know the ID of the dashcard from when we create it.

To add to my suggestion from earlier you could even name the IDs and use them like this:
```
(is (= #{unchanged-question-dashcard-id unchanged-text-dashcard-id}
       (t2/select-pk-set :model/DashboardCard :dashboard_id dashboard-id)))
```

Reply:
> I'm thinking of ways to minimize the number of changes you'd have to make to this test if unrelated parts of the codes change, such as the way text cards are represented.

this makes sense


> (is (= #{unchanged-question-dashcard-id unchanged-text-dashcard-id}
       (t2/select-pk-set :model/DashboardCard :dashboard_id dashboard-id)))
       
This is not possible because dashboardcard are not "restored", when you revert, we create new dashboardcard, so the ids will not match.

Reply:
Ah, of course. You're right.